src/filters/fb_network.rs

```diff
@@ -0,0 +1,314 @@
+            .as_ref()
+            .map(|v| self.builder.create_string(v.as_str()));
+
+        let filter = fb::NetworkFilter::create(
+            &mut self.builder,
+            &fb::NetworkFilterArgs {
+                mask: network_filter.mask.bits(),
+                patterns: patterns,
+                modifier_option: modifier_option,
+                opt_domains: opt_domains,
+                opt_not_domains: opt_not_domains,
+                hostname: hostname,
+                tag: tag,
+                raw_line: raw_line,
+            },
+        );
+
+        self.filters.push(filter);
+        u32::try_from(self.filters.len() - 1).expect("< u32::MAX")
+    }
+
+    pub fn finish(&mut self) -> Vec<u8> {
+        let filters = self.builder.create_vector(&self.filters);
+
+        let unique_domains_hashes = self.builder.create_vector(&self.unique_domains_hashes);
+
+        let storage = fb::NetworkFilterList::create(
+            &mut self.builder,
+            &&fb::NetworkFilterListArgs {
+                network_filters: Some(filters),
+                unique_domains_hashes: Some(unique_domains_hashes),
+            },
+        );
+        self.builder.finish(storage, None);
+
+        let binary = Vec::from(self.builder.finished_data());
+        binary
+    }
+}
+pub struct FlatPatterns<'a> {
+    patterns: Option<flatbuffers::Vector<'a, flatbuffers::ForwardsUOffset<&'a str>>>,
+}
+
+impl<'a> FlatPatterns<'a> {
+    #[inline(always)]
+    pub fn new(
+        patterns: Option<flatbuffers::Vector<'a, flatbuffers::ForwardsUOffset<&'a str>>>,
+    ) -> Self {
+        Self { patterns }
+    }
+
+    #[inline(always)]
+    pub fn iter(&self) -> FlatPatternsIterator {
+        FlatPatternsIterator {
+            patterns: self,
+            len: self.patterns.map_or(0, |d| d.len()),
+            index: 0,
+        }
+    }
+}
+
+pub struct FlatPatternsIterator<'a> {
+    patterns: &'a FlatPatterns<'a>,
+    len: usize,
+    index: usize,
+}
+
+impl<'a> Iterator for FlatPatternsIterator<'a> {
+    type Item = &'a str;
+
+    #[inline(always)]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.patterns.patterns.map_or(None, |fi| {
+            if self.index < self.len {
+                self.index += 1;
+                Some(fi.get(self.index - 1))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl<'a> ExactSizeIterator for FlatPatternsIterator<'a> {
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.len
+    }
+}
+
+pub struct FlatNetworkFilter<'a> {
+    key: u64,
+    owner: &'a NetworkFilterList,
+    fb_filter: &'a fb::NetworkFilter<'a>,
+
+    pub(crate) mask: NetworkFilterMask,
+}
+
+impl<'a> FlatNetworkFilter<'a> {
+    #[inline(always)]
+    pub fn new(
+        filter: &'a fb::NetworkFilter<'a>,
+        index: u32,
+        owner: &'a NetworkFilterList,
+    ) -> Self {
+        let list_address: *const NetworkFilterList = owner as *const NetworkFilterList;
```

Comment:
pointer casting often indicates some problematic design patterns in Rust. I think we can remove it in this instance?
- this `owner` field can increase memory usage since it's an additional address-sized field on each filter, even though they'd all have the same value for all filters in a list
- I only see it used in `NetworkMatchable` to make it conform to the existing API. I'm not opposed to that API changing if needed.

Reply:
`FlatNetworkFilter` is a wrapper over the `fb::NetworkFilter`.  We only create the instance of it when found the the `fb::NetworkFilter` in the `NetworkFilterList` so it does't increase the memory usage because it temporary object on the stack. For now, regex manager is designed to use unique key to find corresponding regex, previously it was an address of the `NetworkFilter`, but now we don't have unique address for every filter because they stored in flatbuffer. To create unique key we use the list address combined with the filter's index. I agree it looks ugly but I don't know how to do it better.

Reply:
Let's understand why we need this unique `key`.
If it's only for regex manager, let's:
1. try to move key calculation here
2. add TODO to make keys unique only in scope of regex manager. I believe once that in one of the next steps we move from multiple flatbuffers to a just one that ables to contain multiple lists. That way we get 1:1 proportion (regex manager: flatbuffer).

Reply:
Also, this could be done in follow up: this PR is already waiting for a long time.

Reply:
`1 regex manager: 1 flatbuffer` definitely sounds like the right way to do it, but agreed it can be followup if needed